src/MISC/fix_eco_force.cpp

```diff
@@ -0,0 +1,642 @@
+  " journal = {Computational Materials Science},\n"
+  " volume = {182},\n"
+  " year = {2020},\n"
+  " pages = {109774},\n"
+  " doi = {j.commatsci.2020.109774},\n"
+  " url = {https://doi.org/10.1016/j.commatsci.2020.109774}\n"
+  "}\n\n";
+
+/* ---------------------------------------------------------------------- */
+
+FixECOForce::FixECOForce(LAMMPS *lmp, int narg, char **arg) :
+    Fix(lmp, narg, arg),
+    dir_filename(NULL), order(NULL), nbr(NULL), list(NULL)
+{
+  if (lmp->citeme) lmp->citeme->add(cite_fix_eco_force);
+
+  // get rank of this processor
+  MPI_Comm_rank(world, &me);
+
+  // check for illegal command
+  if (narg != 7) error->all(FLERR, "Illegal fix eco/force command");
+
+  // set fix flags
+  scalar_flag = 1;          // computes scalar
+  global_freq = 1;          // values can be computed at every timestep
+  extscalar = 1;            // scalar scales with # of atoms
+  peratom_flag = 1;         // quantities are per atom quantities
+  size_peratom_cols = 2;    // # of per atom quantities
+  peratom_freq = 1;         //
+
+  // parse input parameters
+  u_0 = atof(arg[3]);
+  sign = (u_0 >= 0.0 ? 1 : -1);
+  eta = atof(arg[4]);
+  r_cut = atof(arg[5]);
+
+  // read reference orientations from file
+  // work on rank 0 only
+  int n = strlen(arg[6]) + 1;
+  dir_filename = new char[n];
+  strcpy(dir_filename, arg[6]);
+  if (me == 0) {
+    char line[IMGMAX];
+    char *result;
+    int count;
+
+    FILE *infile = fopen(dir_filename, "r");
+    if (infile == NULL) error->one(FLERR, "Fix eco/force file open failed");
+    for (int i = 0; i < 6; ++i) {
+      result = fgets(line, IMGMAX, infile);
+      if (!result) error->one(FLERR, "Fix eco/force file read failed");
+      count = sscanf(line, "%lg %lg %lg", &dir_vec[i][0], &dir_vec[i][1], &dir_vec[i][2]);
+      if (count != 3) error->one(FLERR, "Fix eco/force file read failed");
+    }
+    fclose(infile);
+
+    // calculate reciprocal lattice vectors
+    get_reciprocal();
+
+    squared_cutoff = r_cut * r_cut;
+    inv_squared_cutoff = 1.0 / squared_cutoff;
+    half_u = 0.5 * u_0;
+    inv_eta = 1.0 / eta;
+  }
+
+  // initializations
+  MPI_Bcast(&dir_vec[0][0], 18, MPI_DOUBLE, 0, world);                  // communicate direct lattice vectors
+  MPI_Bcast(&reciprocal_vectors[0][0][0], 18, MPI_DOUBLE, 0, world);    // communicate reciprocal vectors
+  MPI_Bcast(&squared_cutoff, 1, MPI_DOUBLE, 0, world);                  // communicate squared cutoff radius
+  MPI_Bcast(&inv_squared_cutoff, 1, MPI_DOUBLE, 0, world);              // communicate inverse squared cutoff radius
+  MPI_Bcast(&half_u, 1, MPI_DOUBLE, 0, world);                          // communicate half potential energy
+  MPI_Bcast(&inv_eta, 1, MPI_DOUBLE, 0, world);                         // communicate inverse threshold
+
+  // set comm size needed by this Fix
+  if (u_0 != 0.0) comm_forward = 110;
```

Comment:
how are these numbers determined? please add a comment so that if somebody wants to update/refactor the code it is possible to adjust them, if needed.